Generic treectrl cleanup

vadz opened the pull request:
I'm not sure about the last commit, there must have been some reason for introducing `m_hasExplicitFont`, but I just don't see it. If you could please explain why is it needed, it could be reverted. The other 2 should still be applied IMO, but please let me know if you disagree. TIA!

vadz committed:
Use wxBrush objects instead of pointers in wxGenericTreeCtrl

There is no need for double indirection, wxBrush object is already
pointer-like internally.

No real changes, just simplify/optimize the code a little.

vadz committed:
Factor out wxGenericTreeCtrl::InitVisualAttributes()

No real changes, just avoid calling an event handler directly.

vadz committed:
Remove apparently unnecessary m_hasExplicitFont

Just use wxWindow::m_hasFont instead.

paulcor commented on include/wx/generic/treectlg.h:
```diff
@@ -243,8 +244,8 @@ class WXDLLIMPEXP_CORE wxGenericTreeCtrl : public wxTreeCtrlBase,
-    wxBrush             *m_hilightBrush,
-                        *m_hilightUnfocusedBrush;
+    wxBrush              m_hilightBrush,
+                         m_hilightUnfocusedBrush;
```
These members are protected, so user code can get at them. I didn't make this change for fear of breaking something, especially this close to making a new release.

vadz replied:
I didn't even think about this, to be honest, but I don't believe we guarantee compatibility for random, undocumented, protected members (and yes, of course they should have been private in the first place...). Unfortunately it's difficult to prove that nobody really uses them, but it seems like their only occurrences are in the various clones of this file in the other projects, not in the projects actually using these variables in the base class.

paulcor commented on src/generic/treectlg.cpp:
```diff
@@ -1012,18 +1005,16 @@ void wxGenericTreeCtrl::EnableBellOnNoMatch( bool on )
     SetOwnForegroundColour(attr.colFg);
     SetOwnBackgroundColour(attr.colBg);
-    if (!m_hasExplicitFont)
+    if (!m_hasFont)
```
`m_hasFont` will always be true after `SetOwnFont()` is called with a valid font, so subsequent executions of this code may not have the same effect as the first. That's why I had to add `m_hasExplicitFont`.

BTW, why is it necessary to set the font and foreground/background colors here at all? Won't these values be the defaults anyway?

vadz replied:
Oops, yes, of course, thanks for the explanation. A trivial fix would be to add `m_hasFont = false;` after `SetOwnFont()` call because we know that it's not _really_ set. I also don't really understand why do we do this check for the font but not the colours and [the original patch](https://trac.wxwidgets.org/ticket/6738) doesn't really explain it neither...

We do need to set the colours/fonts because we want them to default to those from `GetClassDefaultAttributes()` and not the defaults inherited from the parent.

vadz committed:
Improve wxGenericTreeCtrl colours/fonts updating on theme change

Don't override user-specified colours neither, previously we only did it
for the font for some reason.

Also do override the colours and the font when the theme changes, which
didn't happen after the previous commit because calling SetOwnXXX()
changes m_hasXXX to true, so we need to explicitly reset it back.

vadz commented:
Thinking more about this, it would really make sense to do all this at `wxWindow` level, there is a big part not specific to `wxGeneritcTreeCtrl` at all here... But for now I've just [created a ticket](https://trac.wxwidgets.org/ticket/18826) for this and will just do the minimum change to fix the issue you pointed out (thanks again!).

BTW, `m_normalFont` is also an aberration and shouldn't exist...
